bitcoin/src/address.rs

```diff
@@ -565,22 +621,55 @@ impl<'a> fmt::Display for AddressEncoding<'a> {
 /// * [BIP341 - Taproot: SegWit version 1 spending rules](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki)
 /// * [BIP350 - Bech32m format for v1+ witness addresses](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct Address {
+pub struct Address<V = NetworkValid>
+where
+    V: NetworkValidation,
+{
     /// The type of the address.
     pub payload: Payload,
     /// The network on which this address is usable.
     pub network: Network,
+    /// Marker of the status of network validation.
+    _validation: PhantomData<V>,
 }
+
 #[cfg(feature = "serde")]
 crate::serde_utils::serde_string_impl!(Address, "a Bitcoin address");
```

Comment:
This should be `Address<NetworkUnchecked>`

Reply:
I was unsure how to do this without actually doing it, so here it is if you want to use it.
```rust
macro_rules! serde_string_impl {
    ($name:ident, $expecting:literal) => {
        $crate::serde_utils::serde_string_impl!($name, $expecting, );
    };
    ($name:ident, $expecting:literal, $($gent:ident),*) => {
        impl<'de> $crate::serde::Deserialize<'de> for $name<$($gent),*> {
            fn deserialize<D>(deserializer: D) -> Result<$name<$($gent),*>, D::Error>
            where
                D: $crate::serde::de::Deserializer<'de>,
            {
                use core::fmt::{self, Formatter};
                use core::str::FromStr;

                struct Visitor;
                impl<'de> $crate::serde::de::Visitor<'de> for Visitor {
                    type Value = $name<$($gent),*>;

                    fn expecting(&self, f: &mut Formatter) -> fmt::Result {
                        f.write_str($expecting)
                    }

                    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
                    where
                        E: $crate::serde::de::Error,
                    {
                        let a: $name<$($gent),*> = $name::from_str(v).map_err(E::custom)?;
                        Ok(a)
                    }
                }

                deserializer.deserialize_str(Visitor)
            }
        }

        impl $crate::serde::Serialize for $name {
            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
            where
                S: $crate::serde::Serializer,
            {
                serializer.collect_str(&self)
            }
        }
    };
}
pub(crate) use serde_string_impl;
```
Input from the others please on whether `$gent` is a suitable identifier.

Reply:
Doesn't `$name:ty` work? I think we had to use explicit generics in other macros for some reason but I think this one shouldn't need it.

Reply:
Oh, may well do - I didn't try that.

Reply:
Right, deserialization should be on `Address<NetworkUnchecked>`. Serialization is on both. Done in a6bc605228e010d177a77132cfff6fc3e86fab1d.